blockchain/blockdao/blob_store.go

```diff
@@ -22,4 +36,206 @@ type (
+	}
+	raw, err := bs.kvStore.Get(_blobDataNS, bs.keyForBlock(height))
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get blobs at height %d", height)
+	}
+	blobs, hashes, err := decodeBlob(raw)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to decode blobs at height %d", height)
+	}
+	return blobs, hashes, nil
+}
+
+func (bs *blobStore) PutBlock(blk *block.Block) error {
+	if blk.Height() <= bs.currWriteBlock {
+		return errors.Errorf("block height %d is less than current tip height", blk.Height())
+	}
+	pb := iotextypes.BlobTxSidecars{
+		TxHash:   make([][]byte, 0),
+		Sidecars: make([]*iotextypes.BlobTxSidecar, 0),
+	}
+	for _, act := range blk.Actions {
+		if b := act.BlobTxSidecar(); b != nil {
+			h, err := act.Hash()
+			if err != nil {
+				return err
+			}
+			pb.TxHash = append(pb.TxHash, h[:])
+			pb.Sidecars = append(pb.Sidecars, action.ToProtoSideCar(b))
+		}
+	}
+	raw, err := proto.Marshal(&pb)
+	if err != nil {
+		return errors.Wrapf(err, "failed to put block = %d", blk.Height())
+	}
+	return bs.putBlob(raw, blk.Height(), pb.TxHash)
+}
+
+func (bs *blobStore) putBlob(blob []byte, height uint64, txHash [][]byte) error {
+	var (
+		b              = batch.NewBatch()
+		_, hashes, err = bs.getBlobIndex(height)
+	)
+	switch {
+	case err == nil:
+		// delete old action hash
+		for i := range hashes {
+			b.Delete(_hashSlotNS, hashes[i], "failed to delete hash")
+		}
+	case errors.Cause(err) == db.ErrNotExist:
+		// the slot has not yet been used, do nothing
+	default:
+		return errors.Wrapf(err, "failed to put blob at height %d", height)
+	}
+	// update slot/key to new height/hash mapping
+	var (
+		key   = bs.keyForBlock(height)
+		index = blobIndex{
+			height: height,
+			hashes: txHash,
+		}
+	)
+	b.Put(_hashSlotNS, key, index.serialize(), "failed to put height to hash mapping")
+	for i := range txHash {
+		b.Put(_hashSlotNS, txHash[i], byteutil.Uint64ToBytesBigEndian(height), "failed to put hash to height mapping")
+	}
+	// write the blob data
+	b.Put(_blobDataNS, key, blob, "failed to put blob")
+	if err := bs.kvStore.WriteBatch(b); err != nil {
+		return err
+	}
+	bs.currWriteBlock = height
+	return nil
+}
+
+func decodeBlob(raw []byte) ([]*types.BlobTxSidecar, []string, error) {
+	pb := iotextypes.BlobTxSidecars{}
+	if err := proto.Unmarshal(raw, &pb); err != nil {
+		return nil, nil, err
+	}
+	var (
+		pbBlobs  = pb.GetSidecars()
+		pbHashes = pb.GetTxHash()
+		blobs    = make([]*types.BlobTxSidecar, len(pbBlobs))
+		hashes   = make([]string, len(pbBlobs))
+		err      error
+	)
+	for i := range pbBlobs {
+		blobs[i], err = action.FromProtoBlobTxSideCar(pbBlobs[i])
+		if err != nil {
+			return nil, nil, err
+		}
+		hashes[i] = common.BytesToHash(pbHashes[i]).Hex()
+	}
+	return blobs, hashes, nil
+}
+
+func (bs *blobStore) keyForBlock(b uint64) []byte {
+	b %= bs.totalBlocks
+	return byteutil.Uint64ToBytesBigEndian(b)
+}
```

Comment:
Let's measure the impact of using a ring buffer to make sure it worths the effort and compatible to feature changes or expansions.

Reply:
updated to use block height as storage key